LAB_2

Piotr Suchy 407332

Łączenie z bazą danych przy pomocy skryptu Python i wypisanie przykładowego zapytania:

In [46]:
from sqlalchemy import create_engine
import psycopg2 as pg
import pandas as pd

db_string = "postgresql://wbauer_adb:adb2020@pgsql-196447.vipserv.org:5432/wbauer_adb"

db = create_engine(db_string)

connection_sqlalchemy = db.connect()

In [47]:
# Przykładowe zapytanie: 
# result_set = db.execute("SELECT * FROM city")  
# for r in result_set:  
#     print(r)

1. Ile kategorii filmów mamy w wypożyczalni?

In [48]:
result_set = db.execute("SELECT category_id FROM category")  
for r in result_set:  
    maximum = 0
    if r[0] > maximum:
        maximum = r[0]
print(maximum)

16


2. Wyświetl listę kategorii w kolejności alfabetycznej.


In [49]:
# 2. zapytanie:
result_set = db.execute("SELECT DISTINCT name \
    FROM category \
    ORDER BY name ASC")  
for r in result_set:  
    print(r)

('Action',)
('Animation',)
('Children',)
('Classics',)
('Comedy',)
('Documentary',)
('Drama',)
('Family',)
('Foreign',)
('Games',)
('Horror',)
('Music',)
('New',)
('Sci-Fi',)
('Sports',)
('Travel',)


3. Znajdź najstarszy i najmłodszy film do wypożyczenia.


In [50]:
# 3. zapytanie:
result_set = db.execute("SELECT title, release_year \
    FROM film \
    ORDER BY release_year ASC")  
for r in result_set:
    print(r)

('Chamber Italian', 2006)
('Grosse Wonderful', 2006)
('Airport Pollock', 2006)
('Bright Encounters', 2006)
('Academy Dinosaur', 2006)
('Ace Goldfinger', 2006)
('Adaptation Holes', 2006)
('Affair Prejudice', 2006)
('African Egg', 2006)
('Agent Truman', 2006)
('Airplane Sierra', 2006)
('Alabama Devil', 2006)
('Aladdin Calendar', 2006)
('Alamo Videotape', 2006)
('Alaska Phantom', 2006)
('Date Speed', 2006)
('Ali Forever', 2006)
('Alice Fantasia', 2006)
('Alien Center', 2006)
('Alley Evolution', 2006)
('Alone Trip', 2006)
('Alter Victory', 2006)
('Amadeus Holy', 2006)
('Amelie Hellfighters', 2006)
('American Circus', 2006)
('Amistad Midsummer', 2006)
('Anaconda Confessions', 2006)
('Analyze Hoosiers', 2006)
('Angels Life', 2006)
('Annie Identity', 2006)
('Anonymous Human', 2006)
('Anthem Luke', 2006)
('Antitrust Tomatoes', 2006)
('Anything Savannah', 2006)
('Apache Divine', 2006)
('Apocalypse Flamingos', 2006)
('Apollo Teen', 2006)
('Arabia Dogma', 2006)
('Arachnophobia Rollercoaster', 200

In [51]:
result_set = db.execute("SELECT title, release_year \
    FROM film \
    ORDER BY release_year DESC")  
for r in result_set:
    print(r)

('Chamber Italian', 2006)
('Grosse Wonderful', 2006)
('Airport Pollock', 2006)
('Bright Encounters', 2006)
('Academy Dinosaur', 2006)
('Ace Goldfinger', 2006)
('Adaptation Holes', 2006)
('Affair Prejudice', 2006)
('African Egg', 2006)
('Agent Truman', 2006)
('Airplane Sierra', 2006)
('Alabama Devil', 2006)
('Aladdin Calendar', 2006)
('Alamo Videotape', 2006)
('Alaska Phantom', 2006)
('Date Speed', 2006)
('Ali Forever', 2006)
('Alice Fantasia', 2006)
('Alien Center', 2006)
('Alley Evolution', 2006)
('Alone Trip', 2006)
('Alter Victory', 2006)
('Amadeus Holy', 2006)
('Amelie Hellfighters', 2006)
('American Circus', 2006)
('Amistad Midsummer', 2006)
('Anaconda Confessions', 2006)
('Analyze Hoosiers', 2006)
('Angels Life', 2006)
('Annie Identity', 2006)
('Anonymous Human', 2006)
('Anthem Luke', 2006)
('Antitrust Tomatoes', 2006)
('Anything Savannah', 2006)
('Apache Divine', 2006)
('Apocalypse Flamingos', 2006)
('Apollo Teen', 2006)
('Arabia Dogma', 2006)
('Arachnophobia Rollercoaster', 200

Możemy zauważyć, że układając filmy w kolejności rosnącej i malejącej pod względem daty wydania otrzymujemy ten sam wynik. Dzieje się tak, ponieważ wszystkie te filmy zostały wydane w 2006 roku, przez co niemożliwym jest znalezienie najstarszego i najmłodszego filmu do wypozyczenia.

4. Ile wypożyczeń odbyło się między 2005-07-01 a 2005-08-01?


In [52]:
# 4. zapytanie:
condition = "rental_date > '2005-07-01' AND rental_date < '2005-08-01'"
result_set = db.execute("SELECT rental_date FROM rental WHERE {}".format(condition))  
count = 0
for r in result_set:
    count += 1
print(count)

6709


5. Ile wypożyczeń odbyło się między 2010-01-01 a 2011-02-01?


In [53]:
# 5. zapytanie:
condition = "rental_date > '2010-01-01' AND rental_date < '2011-02-01'"
result_set = db.execute("SELECT rental_date FROM rental WHERE {}".format(condition))  
count = 0
for r in result_set:
    count += 1
print(count)

0


In [54]:
# pomocnicze sprawdzenie: 
# result_set = db.execute("SELECT rental_date FROM rental ORDER BY rental_date DESC")  
# for r in result_set:
#     print(r)

6. Znajdź największą płatność wypożyczenia.


In [55]:
# 6. zapytanie:
result_set = db.execute("SELECT amount \
    FROM payment \
    ORDER BY amount DESC LIMIT 1")  
for r in result_set:
    print(r)

(Decimal('11.99'),)


Największa płatność wypozyczenia to 11.99.

7. Znajdź wszystkich klientów z Polski, Nigerii lub Bangladeszu.


In [56]:
# 7. zapytanie:
condition = "country = 'Poland' OR country = 'Nigeria' OR country = 'Bangladesh'"
result_set = db.execute("SELECT customer_id, first_name, last_name, email, country \
    FROM customer cust \
    JOIN address a ON cust.address_id = a.address_id \
    JOIN city c ON a.city_id = c.city_id \
    JOIN country co ON c.country_id = co.country_id \
    WHERE {} ORDER BY country ASC".format(condition))  
for r in result_set:
    print(r)

(21, 'Michelle', 'Clark', 'michelle.clark@sakilacustomer.org', 'Bangladesh')
(332, 'Stephen', 'Qualls', 'stephen.qualls@sakilacustomer.org', 'Bangladesh')
(329, 'Frank', 'Waggoner', 'frank.waggoner@sakilacustomer.org', 'Bangladesh')
(80, 'Marilyn', 'Ross', 'marilyn.ross@sakilacustomer.org', 'Nigeria')
(198, 'Elsie', 'Kelley', 'elsie.kelley@sakilacustomer.org', 'Nigeria')
(103, 'Gladys', 'Hamilton', 'gladys.hamilton@sakilacustomer.org', 'Nigeria')
(284, 'Sonia', 'Gregory', 'sonia.gregory@sakilacustomer.org', 'Nigeria')
(409, 'Rodney', 'Moeller', 'rodney.moeller@sakilacustomer.org', 'Nigeria')
(286, 'Velma', 'Lucas', 'velma.lucas@sakilacustomer.org', 'Nigeria')
(18, 'Carol', 'Garcia', 'carol.garcia@sakilacustomer.org', 'Nigeria')
(277, 'Olga', 'Jimenez', 'olga.jimenez@sakilacustomer.org', 'Nigeria')
(156, 'Bertha', 'Ferguson', 'bertha.ferguson@sakilacustomer.org', 'Nigeria')
(280, 'Tracey', 'Barrett', 'tracey.barrett@sakilacustomer.org', 'Nigeria')
(250, 'Jo', 'Fowler', 'jo.fowler@sakila

8. Gdzie mieszkają członkowie personelu?


In [57]:
# 8. zapytanie:
result_set = db.execute("SELECT first_name, last_name, address, district, city, country \
    FROM staff s \
    JOIN address a ON s.address_id = a.address_id \
    JOIN city c ON a.city_id = c.city_id \
    JOIN country co ON c.country_id = co.country_id")  
for r in result_set:
    print(r)

('Mike', 'Hillyer', '23 Workhaven Lane', 'Alberta', 'Lethbridge', 'Canada')
('Jon', 'Stephens', '1411 Lillydale Drive', 'QLD', 'Woodridge', 'Australia')


In [58]:
# pomocnicze sprawdzenie:
# result_set = db.execute("SELECT * FROM staff")  
# for r in result_set:
#     print(r)

9. Ilu pracowników mieszka w Argentynie lub Hiszpanii?


In [59]:
# 9. zapytanie:
condition = "country = 'Argentine' OR country = 'Spain'"
result_set = db.execute("SELECT first_name, last_name, address, district, city, country \
    FROM staff s JOIN address a ON s.address_id = a.address_id \
    JOIN city c ON a.city_id = c.city_id \
    JOIN country co ON c.country_id = co.country_id WHERE {}".format(condition))  
q = 0
for r in result_set:
    q += 1
print(q)

0


Żaden z pracowników nie mieszka w Argentynie lub Hiszpani.

10. Jakie kategorie filmów zostały wypożyczone przez klientów?

In [60]:
# 10. zapytanie: 
result_set = db.execute("SELECT DISTINCT name FROM rental r \
    JOIN inventory i ON r.inventory_id = i.inventory_id \
    JOIN film_category f_c ON i.film_id = f_c.film_id \
    JOIN category c ON f_c.category_id = c.category_id")  
for r in result_set:
    print(r)

('Sports',)
('Classics',)
('New',)
('Family',)
('Comedy',)
('Animation',)
('Travel',)
('Music',)
('Drama',)
('Horror',)
('Sci-Fi',)
('Games',)
('Documentary',)
('Foreign',)
('Action',)
('Children',)


Używając SELECT DISTINCT do usunięcia zduplikowanych wierszy otrzymujemy jako wynik listę wszystkich wypożyczonych kategorii filmów. Możemy zauważyć, że wszystkie 16 kategorii zostało wypożyczone.

11. Znajdź wszystkie kategorie filmów wypożyczonych w Ameryce.

In [61]:
# 11. zapytanie:
condition = "country = 'United States'"
result_set = db.execute("SELECT DISTINCT name FROM rental r \
    JOIN inventory i ON r.inventory_id = i.inventory_id \
    JOIN film_category f_c ON i.film_id = f_c.film_id \
    JOIN category c ON f_c.category_id = c.category_id \
    JOIN customer cust ON r.customer_id = r.customer_id \
    JOIN address a ON cust.address_id = a.address_id \
    JOIN city ON a.city_id = city.city_id \
    JOIN country co ON city.country_id = co.country_id WHERE {}".format(condition))  
for r in result_set:
    print(r)

('Sports',)
('Classics',)
('New',)
('Family',)
('Comedy',)
('Animation',)
('Travel',)
('Music',)
('Drama',)
('Horror',)
('Sci-Fi',)
('Games',)
('Documentary',)
('Foreign',)
('Action',)
('Children',)


12. Znajdź wszystkie tytuły filmów, w których grał: Olympia Pfeiffer lub Julia Zellweger lub Ellen Presley


In [62]:
# 12. zapytanie:
condition_1 = "first_name = 'Olympia' AND last_name = 'Pfeiffer'"
condition_2 = "first_name = 'Julia' AND last_name = 'Zellweger'"
condition_3 = "first_name = 'Ellen' AND last_name = 'Presley'"

result_set = db.execute("SELECT DISTINCT title, first_name, last_name \
    FROM film f \
    JOIN film_actor f_a ON f.film_id = f_a.film_id \
    JOIN actor a ON f_a.actor_id = a.actor_id \
    WHERE {0} OR {1} OR {2}".format(condition_1, condition_2, condition_3))  
for r in result_set:
    print(r)

('Badman Dawn', 'Olympia', 'Pfeiffer')
('Bilko Anonymous', 'Ellen', 'Presley')
('Breakfast Goldfinger', 'Julia', 'Zellweger')
('Caribbean Liberty', 'Ellen', 'Presley')
('Casper Dragonfly', 'Ellen', 'Presley')
('Chitty Lock', 'Olympia', 'Pfeiffer')
('Color Philadelphia', 'Olympia', 'Pfeiffer')
('Contact Anonymous', 'Olympia', 'Pfeiffer')
('Cranes Reservoir', 'Julia', 'Zellweger')
('Dares Pluto', 'Julia', 'Zellweger')
('Deep Crusade', 'Olympia', 'Pfeiffer')
('Detective Vision', 'Julia', 'Zellweger')
('Divorce Shining', 'Julia', 'Zellweger')
('Effect Gladiator', 'Olympia', 'Pfeiffer')
('Empire Malkovich', 'Ellen', 'Presley')
('Express Lonely', 'Olympia', 'Pfeiffer')
('Firehouse Vietnam', 'Olympia', 'Pfeiffer')
('Floats Garden', 'Ellen', 'Presley')
('Frogmen Breaking', 'Ellen', 'Presley')
('Fugitive Maguire', 'Olympia', 'Pfeiffer')
('Hanky October', 'Olympia', 'Pfeiffer')
('Hollow Jeopardy', 'Julia', 'Zellweger')
('Homeward Cider', 'Ellen', 'Presley')
('Hyde Doctor', 'Ellen', 'Presley')
('